### PR TITLE
multi-key sort

### LINKED ISSUE #37: SORT with multiple parameters
SORT currently sorts by a single key and allows to reverse the sort direction. 
```
{
  "func": "SORT",
  "holder": "<holderID>",
  "key": "label",
  "reverse": true
}
```

I suggest to either improve SORT with multiple keys and directions or to create a MULTISORT operation like this:

```
{
  "func": "MULTISORT",
  "holder": "<holderID>",
  "sort": [{
     "key":"rank",
    "reverse": true
    },{
     "key":"suite",
    "reverse": true
    }]
}
```

### PR DESCRIPTION
**Pull request by robartsd**
Closes #37, #1135, #1160, and #1163.

Note: some changes made by this PR may change the order of widgets within collections of existing routines if `SELECT` `orderBy` or `SORT` have been used with matching key values.

- [x] move shuffleWidgets() and sortWidgets() to main.js
- [x] add multi-key sort
- [x] add custom order
- [x] add rearrange parameter to SORT
- [x] routine logging summary messages for multi-key
- [x] testcafe
- [x] wiki

## SELECT

* sortBy: allows to sort the collection, can be a property name, key object, or array keys (see _key_ in [SORT](https://github.com/ArnoldSmith86/virtualtabletop/wiki/Functions#sort)) - the collection is sorted AFTER adding elements to it

## SORT

[Tutorial Example](https://virtualtabletop.io/#%7B%0A%20%20%22load%22%3A%20%22https%3A%2F%2Fvirtualtabletop.io%2Flibrary%2FTutorial%2520-%2520Functions%2520-%2520SORT.vtt%22%0A%7D). Press F5/refresh browser after opening room.

This function rearranges widgets, i.e. it changes their stacking order (_z position_) according to the given sort criteria.

Widgets with null values are sorted to the bottom of a pile of widgets, and widgets with identical key values retain their relative stacking order.

If the sorted widgets are not in a pile or holder that arranges its child widgets, the sorted widgets may not appear to change at all - only the stacking order changes. It is recommended to use this operation on holders that auto-align their content.

Parameters:
* holder: _holderID_ (or an array) - specifies the holder that contains the widgets to sort
* collection: _collection_ - specifies the collection of widgets to sort if no holder is specified (defaults to DEFAULT collection)
* key: _property name, key object, or array of property names and/or key objects_ - see below
* reverse: _true/false_ - reverse the order after sorting by key (defaults to false)
* locales: _special_ - see [locales argument](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare#using_locales)
* options: _special_ - see [options argument](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare#using_options)
* rearrange: _true/false_ when set to false the widgets are not rearranged in the room, only the order in the collection changes (only applies when sorting a collection, defaults to true)

### SORT key

In its simplest usage, key simply takes the name of a property as a string. The widgets will be sorted by the value of that property from lowest to highest.

```json
{
  "func": "SORT",
  "holder": "<holderID>",
  "key": "<propertyName>"
}
```

If an array is provided, then each element is treated as a key. Keys are applied from the first array element to the last. Widgets will be sorted based on the first key in the array that returns sort values that are not considered equal (based on locales and options).

```json
{
  "func": "SORT",
  "holder": "<holderID>",
  "key": [
    "<propertyName1>",
    "<propertyName2>"
  ]
}
```

Separate options for each key can be provided by providing an object as a key instead of just a property name. The property name is provided as _key_ within the object. _options_ and _locales_ override any values set in the main function block. Unlike _reverse_ in the main function block, _reverse_ in the key object does not reverse the order of widgets with matching sort values. An array specifying the order that values should be sorted in can be provided in the key object as _order_. Any values found in the _order_ array come before all other values. Key object and property names can be mixed in the same key array.

```json
{
  "func": "SORT",
  "holder": "hand",
  "key": [
    {
      "key": "suit",
      "order": [
        "C",
        "D",
        "S",
        "H",
      ]
    },
   "rankA"
  ]
}
```

Important: The property referenced in _key_ applies a standard string comparison unless the values are actually numbers. So if you want to compare numeric strings, either ensure they are padded with zeros ("02" or "002" rather than "2", because a string value of "11" is considered to be less than "2", as the first character of the string is of lower value) or use the "numeric" option as in the following example (numeric option does not sort negatives before positives):

```json
{
  "func": "SORT",
  "holder": "<holderID>",
  "key": "<propertyName>",
  "reverse": false,
  "options": {
    "numeric": true
  }
}
```

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1158/pr-test (or any other room on that server)

**Comment by robartsd** [comment]
I am having an issue with making sort behave exactly as current implementation when the sort key values are identical. Due to multiple sort keys possibly having different sort order, reverse is applied per key instead of on the sorted array. This is illustrated in the Testcafe test for SELECT and SORT. If the array is reversed before sorting, SELECT passes and SORT fails. If the array is not reversed before sorting SORT passes and SELECT fails. I'm inclined to change the hash for SELECT.

**Review by robartsd — COMMENTED** [review]
I believe this is now compatible with all existing sorts. I'm open to closing #1160 with this unless we want it to wait for legacy servers.

File: client/js/widgets/widget.js [hunk 1]
```diff
@@ -1347,22 +1347,27 @@ export class Widget extends StateManaged {
...
       if(a.func == 'SORT') {
-        setDefaults(a, { key: 'value', reverse: false, collection: 'DEFAULT' });
+        setDefaults(a, { key: 'value', reverse: false, collection: 'DEFAULT', rearrange: true });
         let collection;
-        let reverse = a.reverse ? 'in reverse' : '';
+        let reverse = (a.reverse && !Array.isArray(a.reverse)) ? ' in reverse' : '';
+        let key = asArray(a.key).map((k)=>{
+          if(k !== null && typeof k.key == 'string')
+            return `'${k.key}'`;
+          return typeof k == 'string' ? `'${k}` : k;
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:1356]: I might miss something but it feels like there's a `'` missing somewhere.

**Review by ArnoldSmith86 — COMMENTED** [review]
I'm not sure yet if I want to understand the content of the two functions. I'll have another look tomorrow. The rest looks fine but doesn't really do much except calling the functions. :smile:

File: client/js/widgets/widget.js (hunk 1 above)
Comment (robartsd) [reply client/js/widgets/widget.js:1356]: Yep. Just JSON editor routine debugging output, but still should be fixed.

**Comment by robartsd** [comment]
> I'm not sure yet if I want to understand the content of the two functions. I'll have another look tomorrow. The rest looks fine but doesn't really do much except calling the functions. 😄

`shuffleWidgets()` is exactly the same function that was `Widget.shuffleWidgets()` but moved to `main.js`. `sortWidgets()` has significant changes.

**Comment by 96LawDawg** [comment]
I'm going to try and start testing this.  I read the modified wiki and it generally looks good.  But I don't understand, at all, this paragraph:

"Separate options for each key can be provided by providing an object as a key instead of just a property name. The property name is provided as _key_ within the object. _options_ and _locales_ override the any values set in the main function block. Unlike _reverse_ in the main function block, _reverse_ in the key object does not reverse the order of widgets with matching sort values. An array specifying the order that values should be sorted in can be provided in the key object as _order_. Any values found in the _order_ array come before all other values. Key object and property names can be mixed in the same key array."

**Comment by 96LawDawg** [comment]
I have tested this pretty well I think.  First, I loaded the SORT tutorial in this and on production.  After changing the buttons that did a shuffle, I pressed buttons in multiple combinations in both rooms.  Got the same results in both rooms, which is good evidence that this doesn't break existing behavior (except as noted at the top).

Then, I played around with a couple of rooms that use simple(ish) sorts and everything seemed find.

Finally, I tested each of the parameters a couple of different times, including SELECT (sortBy).  Everything seemed to work.  The only change I think would be to add "rearrange" to the JSON Editor options.  That plus my comment in the previous message about making the wiki easier to understand and maybe making true/false turn blue in JSON, and I think this is good to go from my evaluation.

**Comment by robartsd** [comment]
> Also, as part of that paragraph, I don't understand the role of reverse in the key object. If it doesn't reverse things, what does it do? I took what you had in the default test room and swapped the true and false entries for reverse. Didn't seem to do anything. And if those values are really true and false, then the JSON Editor needs to be tweaked so they show up as blue.

Within the key object `reverse` controls how things that have different values in the key are sorted. At the function block level `reverse` sorts everything, then reverses the result.

```
collection = [
  { 'id": "one", "prop": "b"},
  {"id": "two", "prop": "c"},
  {"id": "three", "prop": "b"}
]

{
  "func": "SORT",
  "key": "prop"
  "reverse": true
}  - sorts to [ "two", "three", "one" ]

{
  "func": "SORT",
  "key": {
     "key": "prop",
     "reverse": true
  }
} - sorts to ["two", "one", "three"]
```

**Comment by 96LawDawg** [comment]
I think I understand that.  But the examples you just posted exist in a simple object, not a more complex one (not sure of the right terms).  So in the example screenshot (from your demo room), the white false and true don't seem to do anything.  Or maybe I still just don't understand.

Screenshot 2022-03-31 212941

**Comment by robartsd** [comment]
> I think I understand that. But the examples you just posted exist in a simple object, not a more complex one (not sure of the right terms). So in the example screenshot (from your demo room), the white false and true don't seem to do anything. Or maybe I still just don't understand.

That was an early idea that was dropped in favor of being able to explicitly reverse any tied sort keys. Sorry to confuse you, I forgot that I had that code in the test room.

The SORT Object button shows the only method supported for sorting only certain keys in reverse. It sorts the hand by suit: Diamonds, Clubs, Hearts, Spades, then by rank within the suits highest first.

```json
    {
      "func": "SORT",
      "holder": "hand",
      "key": [
        {
          "key": "suit",
          "order": [
            "D",
            "C",
            "H",
            "S"
          ]
        },
        {
          "key": "rankA",
          "reverse": true
        }
      ]
    }
```

**Comment by RaphaelAlvez** [comment]
I totally missed this PR.

Why is it still up? Seems pretty finished to me

**Comment by RaphaelAlvez** [comment]
@ArnoldSmith86 there is no way I can code review this
@96LawDawg I would like help with test review since there is so much stuff here

**Comment by 96LawDawg** [comment]
I already tested this back in March and put my comments in back then.  The only thing it needs from my perspective is a change to the JSON Editor to include "rearrange" as one of the properties.

**Comment by robartsd** [comment]
I see that RaphaelAlvez added the requested changes to the JSON editor.

**Review by 96LawDawg — APPROVED** [review]
Approved through test.

**Pull request merged**